client/src/main/java/org/spine3/client/QueryBuilder.java

```diff
@@ -0,0 +1,325 @@
+import com.google.protobuf.FieldMask;
+import com.google.protobuf.Message;
+import org.spine3.base.Identifiers;
+import org.spine3.json.Json;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Collections2.transform;
+import static com.google.common.collect.Sets.newHashSet;
+
+import static org.spine3.protobuf.AnyPacker.unpack;
+
+/**
+ * A builder for the {@link Query} instances.
+ *
+ * <p>The API of this class is inspired by the SQL syntax.
+ *
+ * <p>Calling any of the methods is optional. Call {@link #build() build()} to retrieve
+ * the instance of {@link Query}.
+ *
+ * <p>Calling any of the builder methods overrides the previous call of the given method of
+ * any of its overloads. For example, calling sequentially
+ * <pre>
+ *     {@code
+ *     builder.withMask(mask1)
+ *            .withMask(mask2)
+ *            // optionally some other invocations
+ *            .withMask(mask3)
+ *            .build();
+ *     }
+ * </pre>
+ * is equivalent to calling
+ * <pre>
+ *     {@code
+ *     builder.withMask(mask3)
+ *            .build();
+ *     }
+ * </pre>
+ *
+ * <p>Usage example:
+ * <pre>
+ *     {@code
+ *     final Query query = factory().query()
+ *                                  .select(Customer.class)
+ *                                  .byId(getWestCostCustomersIds())
+ *                                  .withMask("name", "address", "email")
+ *                                  .where({@link QueryParameter#eq eq}("type", "permanent"),
+ *                                         eq("discountPercent", 10),
+ *                                         eq("companySize", Company.Size.SMALL))
+ *                                  .build();
+ *     }
+ * </pre>
+ *
+ * @see QueryFactory#select(Class)
+ */
+public final class QueryBuilder {
+
+    private final QueryFactory queryFactory;
+    private final Class<? extends Message> targetType;
+
+    // All the optional fields are initialized only when and if set
+    // The empty collections make effectively no influence, but null values allow us to create
+    // the query `Target` more efficiently
+
+    @Nullable
+    private Set<?> ids;
+
+    @Nullable
+    private Map<String, Any> columns;
+
+    @Nullable
+    private Set<String> fieldMask;
+
+    QueryBuilder(Class<? extends Message> targetType, QueryFactory queryFactory) {
+        this.targetType = checkNotNull(targetType);
+        this.queryFactory = checkNotNull(queryFactory);
+    }
+
+    /**
+     * Sets the ID predicate to the {@linkplain Query}.
+     *
+     * <p>Though it's not prohibited at compile-time, please make sure to pass instances of the
+     * same type to the argument of this method. Moreover, the instances must be of the type of
+     * the query target type identifier. This method or any of its overload do not check these
```

Comment:
I don't see how a type parameter could help here. A user can anyway call `byId(Collection<Object>)` and there is no adequate way to check the ID types.

I think we may rely on the user here since this restriction corresponds to common sense understanding of how do the queries work.